### PR TITLE
add branches to lake model

### LINKED ISSUE #2537: new zed lake: finish merge command
The new zed lake supports a merge operation that takes a set of tags, reads the commits/objects into a new commit, and outputs the new commit tag.

This task involves implementing the stub in `cmd/zed/lake/merge`.  It is very simple.  Just copy code from `zed lake query` and take stuff out.  It should take 15 minutes to code and a bit more time to write some tests.

### LINKED ISSUE #2626: Git-style branches in Zed lakes
In watching the reactions to the Git-style workflows being built for Zed lakes, a couple of us have thought about extending the metaphor all the way to actually having the concept of "branches". For instance, the Zed lake already provides for the ability to delete by commit to undo mistaken data imports. However, users accustomed to Git workflows might feel even safer working with Zed lakes if they could have the have the sense of "working in a branch". So if they were sitting down to work on shaping or importing a new data source into the lake, by working in such a branch they could do their data adds/deletes off a completely separate journal while feeling there's zero disk that their changes would muck up the "main" production journal that other users/apps are dependent on. If the work in their branch ends up being a failure, they could delete the branch and it'd be as if the data never touched the lake. Or if their work is a success and they want to incorporate the data they'd already shaped and cleaned up in their branch, they could merge it into the main journal.

In thinking this through, @mccanne had the following additional thoughts:

> A branch is just another journal object that points to the main journal and we could pretty easily write the code to  merge a branch journal into the main journal (or into another branch) and the write-conflict logic will be the same as merging commits from staging to main.

Given our need to focus on other priorities, this is feeling like a priority to consider in the MVP1 timeframe.

### LINKED ISSUE #2953: get rid of ScanPools()
lake.Root.ScanPools() should be removed along with the pools API end point but the app currently uses it.  We should coordinate with the app team to use the query endpoint (`from [pools]`) instead of the pools API end point, then we can get rid of this stuff.

### LINKED ISSUE #2975: add commit message to `lake load` endpoint
In the branches PR, we got rid of staging and the API endpoint for `lake load` now generates a dummy commit message because the HTTP body is the posted data.  We could (1) encode the commit message in the URL, (2) include a header containing the marshalled commit message in front of the posted data (yuck), or (3) use mime/multipart?

### PR DESCRIPTION
**Pull request by mccanne**
This commit adds a branch concept to the zed lake.  This required
some rewiring of the snapshot logic to implement merge and a new
approach to locking across two branches for a merge (though the
locking details are not yet implemented).  The TAIL pointer in a
journal now tracks the base pointer in a parent branch.  TAIL is
transactionally guaranteed through locking (TBD).

For now there is only one root branch (main), and branches must
all be descended from this.

"zed lake query" no longer reports "empty log" as it simply
produces no output.

"zed lake ls" no longer lists objects as meta-queries can be used
for this.  Instead, ls lists pools and branches.

The lake commands "add", "commit", and "status" are deprecated.
The lake command "load" does an automatic commit in the backend.

Closes #2626
Closes #2537
Closes #2975

### CODE REVIEW

**Comment by mccanne** [comment]
I still need to update lake design.md... will work on that in the morning but wanted to get this up!

File: lake/branch.go [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+func (b *BranchConfig) Path(poolPath *storage.URI) *storage.URI {
+	return poolPath.AppendPath(b.ID.String())
+}
+
+func (b *BranchConfig) Create(ctx context.Context, engine storage.Engine, poolPath *storage.URI, order order.Which, base journal.ID) error {
+	_, err := commit.Create(ctx, engine, b.Path(poolPath), order, base)
+	return err
+}
+
+func (b *BranchConfig) Open(ctx context.Context, engine storage.Engine, parent *storage.URI, pool *Pool) (*Branch, error) {
```
Comment (mattnibs) [inline lake/branch.go:57]: I think poolPath makes this clearer, and is consistent with the Create call:
```suggestion
func (b *BranchConfig) Open(ctx context.Context, engine storage.Engine, poolPath *storage.URI, pool *Pool) (*Branch, error) {
```

File: lake/pool.go [hunk 2]
```diff
@@ -67,383 +59,181 @@ func (p *PoolConfig) Path(root *storage.URI) *storage.URI {
...
+	poolPath := p.Path(root)
+	// branchesPath is the path to the kvs journal of BranchConfigs
+	// for the pool while the commit log is stored in <pool-id>/<branch-id>.
+	branchesPath := poolPath.AppendPath(BranchesTag)
+	// create the branches journal store
+	_, err := kvs.Create(ctx, engine, branchesPath, branchStoreTypes)
+	if err != nil {
+		return err
+	}
+	// create the main branche in the branches journal store
```
Comment (mattnibs) [inline lake/pool.go:74]:
```suggestion
	// create the main branch in the branches journal store
```

File: lake/branch.go [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+	log    *commit.Log
+	base   commit.View
+}
+
+func newBranchConfig(name string, parent ksuid.KSUID) *BranchConfig {
+	//XXX need to take parent args and check for a cycle.
+	// also check that parent journal ID exists in parent.
+	return &BranchConfig{
+		Name:   name,
+		ID:     ksuid.New(),
```
Comment (mattnibs) [inline lake/branch.go:43]: I'm kind of wondering why even bother having an ID for branches. I was moving through the zedlake fs and found it harder to follow what was going on with the ids for branches. I get why we did this for pool names- we didn't want to break links when pools were renamed. This feels like less of a problem with branches.

File: lake/branch.go [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+		Parent: parent,
```
Comment (mattnibs) [inline lake/branch.go:44]: I don't grasp all the implications of doing this... but: It feels limiting to have a branch point to a parent branch vs doing what git does and having a branch to start at a particular commit. If we followed the way you have it coded up here, would you not be able to delete a branch that has child branches? What would happen to the child branches if their parent was deleted?

I suppose this would require making it so each commit referenced what its parent commit is. It would also mean that you'd have to keep commits centralized in pool... Did I mention I don't follow grasp the implications of this suggestion!?

File: service/ztests/add-empty.yaml [hunk 5]
```diff
@@ -1,7 +1,7 @@
 script: |
   source service.sh
   zapi create -q -p test
-  zapi add -q -p test -
+  zapi load -q -p test -
```
Comment (mattnibs) [inline service/ztests/add-empty.yaml:4]: Rename this file load-empty

File: service/ztests/add-garbage.yaml [hunk 6]
```diff
@@ -1,7 +1,7 @@
 script: |
   source service.sh
   zapi create -q -p test
-  zapi add -q -p test -
```
Comment (mattnibs) [inline service/ztests/add-garbage.yaml:4]: Rename this file load-garbage?

File: service/core.go [hunk 7]
```diff
@@ -138,16 +138,15 @@ func (c *Core) addAPIServerRoutes() {
...
-	c.authhandle("/pool/{pool}/add", handleAdd).Methods("POST")
-	c.authhandle("/pool/{pool}/delete", handleDelete).Methods("POST")
-	c.authhandle("/pool/{pool}/squash", handleSquash).Methods("POST")
-	c.authhandle("/pool/{pool}/staging", handleScanStaging).Methods("GET")
-	c.authhandle("/pool/{pool}/staging/{commit}", handleCommit).Methods("POST")
+	c.authhandle("/pool/{pool}/{branch}", handleBranchDelete).Methods("DELETE")
+	c.authhandle("/pool/{pool}", handleBranchPost).Methods("POST")
+	c.authhandle("/pool/{pool}/{branch}", handleBranchLoad).Methods("POST")
+	c.authhandle("/pool/{pool}/{branch}/merge", handleBranchMerge).Methods("POST")
+	c.authhandle("/pool/{pool}/{branch}/delete", handleDelete).Methods("POST")
```
Comment (mattnibs) [inline service/core.go:149]: Usually rest has the object name before id. Maybe put the word `branch` after the poolId and before branchId?

File: lake/branch.go (hunk 4 above)
Comment (mccanne) [reply lake/branch.go:44]: Good point!

File: service/ztests/curl-stats.yaml [hunk 8]
```diff
@@ -1,10 +1,9 @@
 script: |
   source service.sh
   curl -X POST -d '{"name":"test"}' http://$ZED_LAKE_HOST/pool > pool.json
-  poolID=$(zq -f text 'cut value.id' pool.json)
-  curl -X POST -d @- http://$ZED_LAKE_HOST/pool/$poolID/add > add.json
-  commitID=$(zq -f text 'cut value.commit' add.json)
-  curl -X POST -d '{}' http://$ZED_LAKE_HOST/pool/$poolID/staging/$commitID
+  poolID=$(zq -f text 'cut value.pool.id' pool.json)
+  branchID=$(zq -f text 'cut value.branch.id' pool.json)
+  curl -X POST -d @- http://$ZED_LAKE_HOST/pool/$poolID/$branchID > load.json
```
[the remark this answers is not in the document]
Comment (mccanne) [reply service/ztests/curl-stats.yaml:6]: But these two lines post the data into the branch to create the stats...?

[the remark this answers is not in the document]
Comment (mccanne) [reply service/ztests/curl-stats.yaml:6]: `-d @-` is the cryptic option...

File: api/client/connection.go [hunk 9]
```diff
@@ -249,16 +249,30 @@ func (c *Connection) PoolRemove(ctx context.Context, id ksuid.KSUID) error {
...
-		t[i] = tag.String()
+func (c *Connection) BranchPost(ctx context.Context, poolID ksuid.KSUID, payload api.BranchPostRequest) (*Response, error) {
+	req := c.Request(ctx).
+		SetBody(payload)
+	req.Method = http.MethodPost
+	req.URL = path.Join("/pool", poolID.String())
+	resp, err := c.stream(req)
+	var errRes *ErrorResponse
+	if errors.As(err, &errRes) && errRes.StatusCode() == http.StatusConflict {
+		return nil, ErrPoolExists
```
[the remark this answers is not in the document]
Comment (mccanne) [reply api/client/connection.go:260]: adding ErrBranchExists

File: api/client/connection.go [hunk 10]
```diff
@@ -249,16 +249,30 @@ func (c *Connection) PoolRemove(ctx context.Context, id ksuid.KSUID) error {
...
-	req.Method = http.MethodGet
-	req.URL = path.Join("/pool", pool.String(), "staging")
-	return c.stream(req)
+		SetBody(api.BranchMergeRequest{at.String()})
+	req.Method = http.MethodPost
+	req.URL = path.Join("/pool", poolID.String(), branchID.String(), "merge")
+	resp, err := c.stream(req)
+	var errRes *ErrorResponse
+	if errors.As(err, &errRes) && errRes.StatusCode() == http.StatusConflict {
+		return nil, ErrPoolExists
```
[the remark this answers is not in the document]
Comment (mccanne) [reply api/client/connection.go:273]: Deleting as this doesn't happen for a merge.

File: api/client/connection.go [hunk 11]
```diff
@@ -208,13 +206,15 @@ func (c *Connection) PoolGet(ctx context.Context, id ksuid.KSUID) (*Response, er
...
 	req := c.Request(ctx)
-	req.Method = http.MethodGet
-	req.URL = path.Join("/pool", id.String(), "stats")
+	req.Method = http.MethodPost
+	req.URL = "/ids"
+	req.SetBody(api.IDsRequest{poolName, branchName})
 	r, err := c.stream(req)
 	var errRes *ErrorResponse
 	if errors.As(err, &errRes) && errRes.StatusCode() == http.StatusNotFound {
+		//XXX
```
[the remark this answers is not in the document]
Comment (mccanne) [reply api/client/connection.go:217]: deleting

File: service/ztests/curl-stats.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (nwt) [reply service/ztests/curl-stats.yaml:6]: 🤦

File: lake/journal/queue.go [hunk 12]
```diff
@@ -147,12 +155,16 @@ func (q *Queue) Open(ctx context.Context, head, tail ID) (io.Reader, error) {
 			return nil, err
 		}
 		if head == Nil {
-			return nil, ErrEmpty
+			// Return an empty reader when the journal is empty.
+			// This is preferred over returning ErrEmpty and
+			// havings layers above report the error message instead
+			// of simply processing an empty input without error.
+			return strings.NewReader(""), nil
```
Comment (mattnibs) [inline lake/journal/queue.go:162]: 👍🏻

File: api/client/connection.go [hunk 13]
```diff
@@ -31,6 +30,8 @@ var (
 	ErrPoolNotFound = errors.New("pool not found")
 	// ErrPoolExists returns when specified the pool already exists.
 	ErrPoolExists = errors.New("pool exists")
+	// ErrBranchExists returns when specified the branch already exists.
```
Comment (nwt) [inline api/client/connection.go:33]:
```suggestion
	// ErrBranchExists is returned when the specified branch already exists.
```

**Review by mattnibs — APPROVED** [review]

File: cmd/zed/lake/ls/command.go [hunk 14]
```diff
@@ -64,10 +65,13 @@ func (c *Command) Run(args []string) error {
 		return err
 	}
 	var query string
-	if c.lakeFlags.PoolName == "" {
+	if poolName == "" {
 		query = "from [pools]"
 	} else {
-		query = fmt.Sprintf("from '%s'[objects]", c.lakeFlags.PoolName)
+		if strings.IndexByte(poolName, '\'') >= 0 {
```
[the remark this answers is not in the document]
Comment (mccanne) [reply cmd/zed/lake/ls/command.go:71]: At some point we should define the legal syntax for a pool/branch name.  For now, this is just protecting against a printf-d syntax error regarding the single quote so I will leave it as is.

File: lake/branch.go (hunk 3 above)
Comment (mccanne) [reply lake/branch.go:43]: Good point but as we discussed... leaving fo now

File: lake/branch.go (hunk 4 above)
Comment (mccanne) [reply lake/branch.go:44]: Branches v2 coming soon!

File: service/core.go (hunk 7 above)
Comment (mccanne) [reply service/core.go:149]: But the resource is named `/pool/poolID/branchID` then we are applying a verb to that resource, which is not exactly REST-y but I don't see how swapping it makes more REST-y sense, i.e., applying say the verb `delete` to the pool with an branchID under that...?

File: api/client/connection.go (hunk 13 above)
Comment (mccanne) [reply api/client/connection.go:33]: caught the others there too...

**Comment by mccanne** [comment]
Ok, I'll hold off on merging until app integration tests pass...

File: lake/api/api.go [hunk 15]
```diff
@@ -15,28 +17,110 @@ import (
...
+		return nil, fmt.Errorf("internal error: multiple pools found with same name: %s", name)
+	}
+}
+
+func LookupPoolByID(ctx context.Context, api Interface, id ksuid.KSUID) (*lake.PoolConfig, error) {
+	d := newQueryDriver(lake.PoolConfig{})
+	zed := fmt.Sprintf("from [pools] | id == from_hex('%s')", idToHex(id))
+	_, err := api.Query(ctx, d, zed)
+	if err != nil {
+		fmt.Println("XERR", err)
```
Comment (nwt) [inline lake/api/api.go:65]:
```suggestion
```

File: lake/api/driver.go [hunk 16]
```diff
@@ -0,0 +1,41 @@
+package api
+
+import (
+	"github.com/brimdata/zed/api"
+	"github.com/brimdata/zed/zbuf"
+	"github.com/brimdata/zed/zson"
+)
+
+type queryDriver struct {
```
Comment (nwt) [inline lake/api/driver.go:9]: Nit: Call this file `querydriver.go` to match this.

File: lake/api/driver.go [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+		d.results = append(d.results, v)
+	}
+	return nil
+}
+
+func (*queryDriver) Warn(msg string) error {
+	return nil
+}
+
+func (*queryDriver) ChannelEnd(channelID int) error {
```
Comment (nwt) [inline lake/api/driver.go:35]:
```suggestion
func (*queryDriver) Warn(string) error {
	return nil
}

func (*queryDriver) ChannelEnd(int) error {
```

File: cmd/zed/lake/load/command.go [hunk 18]
```diff
@@ -21,11 +21,14 @@ import (
...
 var Load = &charm.Spec{
 	Name:  "load",
-	Usage: "load [options] file|S3-object|- ...",
+	Usage: "load [options] -p pool[/branch] file|S3-object|- ...",
 	Short: "add and commit data to a pool",
 	Long: `
 The load command adds data to a pool and commits it in one automatic operation.
 See documentation on "zed lake add" and "zed lake commit" for more details.
+
+A branch must be specified with the -p option.  If a pool name is given
```
[the remark this answers is not in the document]
Comment (mccanne) [reply cmd/zed/lake/load/command.go:30]: "pool and branch"

File: lake/api/api.go (hunk 15 above)
Comment (mccanne) [reply lake/api/api.go:65]: oopsies

File: service/core.go (hunk 7 above)
Comment (nwt) [reply service/core.go:149]: @mccanne: I think the suggestion is to change `/pool/{pool}/{branch}` to `/pool/{pool}/branch/{branch}`. (I'm in favor.)

File: service/ztests/curl-pool-get.yaml [hunk 19]
```diff
@@ -1,12 +1,15 @@
+skip: GET /pool/* is deprecated
```
Comment (nwt) [inline service/ztests/curl-pool-get.yaml:1]: Revert the changes here and keep this test until we kill `GET /pool` and `GET /pool/{pool}`?

File: service/handlers_test.go [hunk 20]
```diff
@@ -322,6 +314,7 @@ func TestRequestID(t *testing.T) {
 		assert.NotEqual(t, "", res2.Header().Get("X-Request-ID"))
 	})
 	t.Run("PropagatesID", func(t *testing.T) {
+		t.Skip("GET /pool is deprecated")
```
Comment (nwt) [inline service/handlers_test.go:317]:
```suggestion
```

File: service/handlers_test.go [hunk 21]
```diff
@@ -313,6 +304,7 @@ func TestNoEndSlashSupport(t *testing.T) {
 func TestRequestID(t *testing.T) {
 	ctx := context.Background()
 	t.Run("GeneratesUniqueID", func(t *testing.T) {
+		t.Skip("GET /pool is deprecated")
```
Comment (nwt) [inline service/handlers_test.go:307]:
```suggestion
```

File: service/handlers_test.go [hunk 22]
```diff
@@ -246,9 +237,9 @@ func TestPoolStatsNoData(t *testing.T) {
 
 func TestPoolPostNameOnly(t *testing.T) {
 	_, conn := newCore(t)
-	pool := conn.TestPoolPost(api.PoolPostRequest{Name: "test"})
-	assert.Equal(t, "test", pool.Name)
-	assert.NotEqual(t, "", pool.ID)
+	poolID, _ := conn.TestPoolPost(api.PoolPostRequest{Name: "test"})
+	//assert.Equal(t, "test", pool.Name)
```
Comment (nwt) [inline service/handlers_test.go:241]:
```suggestion
```

File: lake/api/remote.go [hunk 23]
```diff
@@ -155,49 +166,15 @@ func unmarshal(res *client.Response, i interface{}) error {
...
-func (r *RemoteSession) Delete(ctx context.Context, poolID ksuid.KSUID, tags []ksuid.KSUID, commit *api.CommitRequest) (ksuid.KSUID, error) {
-	res, err := r.conn.Delete(ctx, poolID, tags)
+func (r *RemoteSession) Delete(ctx context.Context, poolID, branchID ksuid.KSUID, tags []ksuid.KSUID, commit *api.CommitRequest) (ksuid.KSUID, error) {
+	res, err := r.conn.Delete(ctx, poolID, branchID, tags)
 	if err != nil {
 		return ksuid.Nil, err
 	}
 	defer res.Body.Close()
-	var staged api.StagedCommit
+	var staged api.CommitResponse
```
Comment (nwt) [inline lake/api/remote.go:175]: Nit: `staged` no longer seems like the right name for this.

File: lake/commit/snapshot.go [hunk 24]
```diff
@@ -19,6 +18,7 @@ var (
 type View interface {
 	Lookup(ksuid.KSUID) (*segment.Reference, error)
 	Select(extent.Span, order.Which) Segments
+	SelectAll() Segments
```
Comment (nwt) [inline lake/commit/snapshot.go:21]: Doesn't have to happen now but `SelectAll` should go away since it's equivalent to `Select` with a nil `extent.Span`.

File: lake/mock/mock.go [hunk 25]
```diff
@@ -46,9 +48,11 @@ func NewPool(s string) (Pool, error) {
 		}
 	}
 	return Pool{
-		name:   name,
-		id:     fakeID(name),
-		layout: layout,
+		name:     name,
+		id:       fakeID(name),
+		branch:   branchName,
+		branchID: fakeID(branchName),
```
Comment (nwt) [inline lake/mock/mock.go:54]: Nit: Initialize fields in the order they're delcared.
```suggestion
		name:     name,
		branch:   branchName,
		id:       fakeID(name),
		branchID: fakeID(branchName),
```

File: service/core.go [hunk 26]
```diff
@@ -138,21 +138,22 @@ func (c *Core) addAPIServerRoutes() {
...
```
Comment (nwt) [inline service/core.go:149]: Nit: Keep these sorted by path.

**Review by nwt — APPROVED** [review]

File: lake/branch.go (hunk 3 above)
Comment (mccanne) [reply lake/branch.go:43]: They will go away in branches v2.

File: service/core.go (hunk 7 above)
Comment (mattnibs) [reply service/core.go]: What @nwt said
